prql-compiler/src/semantic/resolver.rs

```diff
@@ -631,8 +655,27 @@ impl Resolver {
         Ok(r)
     }
 
-    fn cast_built_in_function(&mut self, closure: Closure) -> Result<Result<Expr, Closure>> {
-        let is_std = closure.name.as_ref().map(|n| n.path.as_slice() == ["std"]);
+    fn fold_function_types(&mut self, mut closure: Func) -> Result<Func> {
+        closure.params = closure
+            .params
+            .into_iter()
+            .map(|p| -> Result<_> {
+                Ok(FuncParam {
+                    ty: self.fold_ty_or_expr(p.ty)?,
+                    ..p
+                })
+            })
+            .try_collect()?;
+        closure.return_ty = self.fold_ty_or_expr(closure.return_ty)?;
+        Ok(closure)
+    }
+
+    fn cast_built_in_function(&mut self, closure: Func) -> Result<Result<Expr, Func>> {
```

Comment:
If we can add notes to signatures with things like double `Result`s, that would be great — I do think signatures are a convenient abstraction, and when those are ambiguous it becomes harder to understand (I realize it's not a new func though, and we should balance the docstrings against slowing you down...)